client/src/components/Header/components/language-list.tsx

```diff
@@ -0,0 +1,166 @@
+      return;
+    }
+    setShowList(true);
+  };
+
+  const handleLanguageChange = (
+    event: React.MouseEvent<HTMLButtonElement>
+  ): void => {
+    const selectedLanguage = event.currentTarget.dataset.value;
+    const isSelecetedCurrentLanguage =
+      selectedLanguage === clientLocale || selectedLanguage === undefined;
+
+    event.preventDefault();
+    listButtonRef.current?.focus();
+    setShowList(false);
+    if (isSelecetedCurrentLanguage) {
+      return;
+    }
+    const path = createLanguageRedirect({
+      clientLocale,
+      lang: selectedLanguage
+    });
+    if (navigate) {
+      return navigate(path);
+    }
+  };
+
+  const handleMenuKeyDown = (
+    event: React.KeyboardEvent<HTMLButtonElement | HTMLAnchorElement>
+  ) => {
+    if (event.key === 'Escape') {
+      listButtonRef.current?.focus();
+      setShowList(false);
+      event.preventDefault();
+    }
+  };
+
+  const handleBlur = (event: React.FocusEvent<HTMLButtonElement>) => {
+    if (
+      event.relatedTarget &&
+      !event.relatedTarget.closest('.nav-list') &&
+      event.relatedTarget !== listButtonRef.current
+    ) {
+      setShowList(false);
+    }
+  };
+
+  const handleLastLangaugeKeys = (
+    event: React.KeyboardEvent<HTMLButtonElement | HTMLAnchorElement>
+  ) => {
+    const DoKeyPress = new Map<string, { select: () => void }>([
+      [
+        'Escape',
+        {
+          select: () => {
+            listButtonRef.current?.focus();
+            setShowList(false);
+            event.preventDefault();
+          }
+        }
+      ],
+      [
+        'Tab',
+        {
+          select: () => {
+            const camperPressedTheShiftKey = event.shiftKey;
+            if (!camperPressedTheShiftKey) {
+              setShowList(false);
+            }
+          }
+        }
+      ]
+    ]);
+    DoKeyPress.get(event.key)?.select();
+  };
+
+  const getHandleLanguageKeys = (languagePosition: number) => {
+    const lastLanguage = locales.length - 1;
+    if (languagePosition === lastLanguage) {
+      return handleLastLangaugeKeys;
+    } else handleMenuKeyDown;
+  };
+  return (
+    <>
+      <button
+        id='toggle-button-nav'
+        className='lang-button-nav'
+        title={t('buttons.change-language')}
+        aria-label={t('buttons.change-language')}
+        aria-controls='nav-lang-list'
+        aria-expanded={showList}
+        ref={listButtonRef}
+        onBlur={handleBlur}
+        onClick={handleClick}
+        onKeyDown={handleMenuKeyDown}
+      >
+        <LanguageGlobe />
+      </button>
+      <ul
+        id='nav-lang-list'
+        className='nav-list'
+        aria-labelledby='toggle-button-nav'
+      >
+        {locales.map((lang, index) => (
+          <li key={'lang-' + lang}>
+            <button
+              className='nav-link nav-lang-list-option'
+              data-value={lang}
+              onClick={handleLanguageChange}
+              onKeyDown={getHandleLanguageKeys(index)}
```

Comment:
```suggestion
              onKeyDown={event => getHandleLanguageKeys(event, index)}
```

If we want to pass `index` into the handler I think we need to call it like this. And then adjust `getHandleLanguageKeys` appropriately by adding the event to the parameter list and then passing the event into the functions it calls.

Reply:
I don't think this is needed, because the function will return `handleLastLangaugeKeys` or `handleMenuKeyDown`, which should act like `handleLanguageChange` in the `onClick` attribute.

As far as I know if event handler is used in React, React will handle passing the event if you write `handleLanguageChange`, it will read it as `event => handleLanguageChange(event)`. 🤔

Although it isn't necessary as far as I know, it's a good call, because I can see it helping while refactoring this in the header PR.